### PR TITLE
Feature/download component

### PR DESCRIPTION
**Pull request by bkheadwire**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | [[RTC] Download Component](https://github.com/adobe/aem-core-wcm-components/issues/217)
| Patch: Bug Fix?          | N/A?
| Minor: New Feature?      | Yes
| Major: Breaking Change?  | No
| Tests Added + Pass?      | Only minimally enough to get it to build
| Documentation Provided   | No readme yet but some comments (mostly javadocs for the model)
| Any Dependency Changes?  | No
| License                  | Apache License, Version 2.0

Was asked to make a PR for this implementation of the download component so the core team can take a look at it. It's definitely still a WIP.

### CODE REVIEW

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/DownloadImpl.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
+package com.adobe.cq.wcm.core.components.internal.models.v1;
+
+import com.adobe.cq.export.json.ComponentExporter;
+import com.adobe.cq.export.json.ExporterConstants;
+import com.adobe.cq.wcm.core.components.models.Download;
+import com.day.cq.commons.DownloadResource;
+import com.day.cq.dam.api.Asset;
+import com.day.cq.wcm.api.designer.Style;
+import com.day.crx.JcrConstants;
```
Comment (kwin) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/DownloadImpl.java]: Wrong package? Rather use constant from Javax.jcr.Property or from https://helpx.adobe.com/experience-manager/6-4/sites/developing/using/reference-materials/javadoc/com/day/cq/commons/jcr/JcrConstants.html

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/DownloadImpl.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+@Model(adaptables = SlingHttpServletRequest.class,
+       adapters = {Download.class, ComponentExporter.class},
+       resourceType = DownloadImpl.RESOURCE_TYPE)
+@Exporter(name = ExporterConstants.SLING_MODEL_EXPORTER_NAME, extensions = ExporterConstants.SLING_MODEL_EXTENSION)
+public class DownloadImpl implements Download {
+
+    private static final Logger LOG = LoggerFactory.getLogger(DownloadImpl.class);
+
+    public final static String RESOURCE_TYPE = "mnrdlm/brand/components/content/download";
```
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/DownloadImpl.java]: Should be: `core/wcm/components/download/v1/download`

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/DownloadImpl.java [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+                    }
+                }
+            }
+        }
+    }
+
+    @Nonnull
+    @Override
+    public String getExportedType() {
+        return request.getResource().getResourceType();
```
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/DownloadImpl.java]: I'd use `resource.getResourceType()`.

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Download.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+import com.adobe.cq.export.json.ComponentExporter;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Sling model for the Download component.
+ */
+public interface Download extends ComponentExporter {
+
+    String PN_TITLE_FROM_ASSET = "titleFromAsset";
```
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Download.java]: Same as above, I'd call it `useTitleFromAsset`.

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Download.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    String PN_DESCRIPTION_FROM_ASSET = "descriptionFromAsset";
```
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Download.java]: Same as above, I'd call it `useDescriptionFromAsset`.

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Download.java (hunk 4 above)
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Download.java]: Please add javadocs.

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Download.java (hunk 5 above)
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Download.java]: Please add javadocs.

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/DownloadImpl.java [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+    private long lastModified = 0;
+
+    @PostConstruct
+    protected void initModel() {
+        String fileReference = properties.get(DownloadResource.PN_REFERENCE, String.class);
+        titleFromAsset = properties.get(PN_TITLE_FROM_ASSET, titleFromAsset);
+        descriptionFromAsset = properties.get(PN_DESCRIPTION_FROM_ASSET, descriptionFromAsset);
+        if (currentStyle != null) {
+            if (StringUtils.isBlank(ctaText)) {
+                ctaText = currentStyle.get("ctaText", String.class);
```
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/DownloadImpl.java]: we should define an API constant for `ctaText`.

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/DownloadImpl.java [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+            }
+            titleType = currentStyle.get("titleType", String.class);
```
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/DownloadImpl.java]: we should define an API constant for `titleType`.

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/DownloadImpl.java [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+                        lastModified = assetLastModified;
+                    }
+
+                    downloadUrl = downloadAsset.getPath();
+
+                    StringBuilder imagePathBuilder = new StringBuilder();
+
+                    String resourcePath = resourceResolver.map(request, resource.getPath());
+
+                    imagePathBuilder.append(resourcePath).append(".coreimg.jpeg");
```
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/DownloadImpl.java]: we should define a constant for the string ".coreimg.jpeg".

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/DownloadImpl.java [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+                    if (lastModified > 0) {
+                        imagePathBuilder.append("/").append(lastModified).append(".jpeg");
```
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/DownloadImpl.java]: we should define a constant for the string ".jpeg".

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/DownloadImpl.java [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+                    }
+
+                    imagePath = imagePathBuilder.toString();
+
+                    if (titleFromAsset) {
+                        String assetTitle = downloadAsset.getMetadataValue("dc:title");
```
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/DownloadImpl.java]: we should define a constant for the string "dc:title".

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/DownloadImpl.java [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+                        if (StringUtils.isNotBlank(assetTitle)) {
+                            title = assetTitle;
+                        }
+                    }
+                    if (descriptionFromAsset) {
+                        String assetDescription = downloadAsset.getMetadataValue("dc:description");
```
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/DownloadImpl.java]: we should define a constant for the string "dc:description".

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Download.java [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+    default String getDescription(){
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * Returns the url to the asset.
+     *
+     * @return the asset url
+     */
+    default String getDownloadUrl(){
```
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Download.java]: I'd call it `getUrl()` to be consistent with other method names like `getTitle`, `getDescription`.

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Download.java [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+     *
+     * @return the path to the image representation of the asset.
+     */
+    default String getImagePath()
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * Returns the element to be used for the title as defined in the component policy.
```
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Download.java]: I would specify: `the HTML element to be used...`

File: bundles/core/src/test/java/com/adobe/cq/wcm/core/components/internal/models/v1/DownloadImplTest.java [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+        assertEquals(DAM_TITLE, download.getTitle());
+        assertEquals(DAM_DESCRIPTION, download.getDescription());
+        assertEquals(PDF_ASSET_PATH, download.getDownloadUrl());
+    }
+
+    private Download getDownloadUnderTest(String resourcePath) {
+        return getDownloadUnderTest(resourcePath, null);
+    }
+
+    private Download getDownloadUnderTest(String resourcePath, Style currentStyle) {
```
Comment (jckautzmann) [inline bundles/core/src/test/java/com/adobe/cq/wcm/core/components/internal/models/v1/DownloadImplTest.java]: This block could be simplified: see for example: https://github.com/adobe/aem-core-wcm-components/blob/development/bundles/core/src/test/java/com/adobe/cq/wcm/core/components/internal/models/v1/CarouselImplTest.java#L82

File: content/src/content/jcr_root/apps/core/wcm/components/download/v1/download/.content.xml [hunk 15]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<jcr:root xmlns:sling="http://sling.apache.org/jcr/sling/1.0" xmlns:cq="http://www.day.com/jcr/cq/1.0" xmlns:jcr="http://www.jcp.org/jcr/1.0"
+    jcr:description="Creates a tile for a downloadable dam asset"
```
Comment (jckautzmann) [inline content/src/content/jcr_root/apps/core/wcm/components/download/v1/download/.content.xml]: DAM should be upper case.

File: content/src/content/jcr_root/apps/core/wcm/components/download/v1/download/_cq_dialog/.content.xml [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+                                                sling:resourceType="granite/ui/components/coral/foundation/well"
+                                                granite:hide="${cqDesign.titleHidden}">
+                                                <items jcr:primaryType="nt:unstructured">
+                                                    <title
+                                                        jcr:primaryType="nt:unstructured"
+                                                        sling:resourceType="granite/ui/components/coral/foundation/form/textfield"
+                                                        fieldDescription="A title to display as the headline for the download."
+                                                        fieldLabel="Title"
+                                                        name="./jcr:title"
+                                                        value="${cqDesign._jcr_description}"/>
```
Comment (jckautzmann) [inline content/src/content/jcr_root/apps/core/wcm/components/download/v1/download/_cq_dialog/.content.xml]: Shouldn't it be `cqDesign._jcr_title`?

File: content/src/content/jcr_root/apps/core/wcm/components/download/v1/download/clientlibs/editor/js/download.js [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+
+    $(document).on("dialog-loaded", function(e) {
+        var $dialog = e.dialog;
+        var $dialogContent = $dialog.find(dialogContentSelector);
+        var dialogContent = $dialogContent.length > 0 ? $dialogContent[0] : undefined;
+
+        if (dialogContent) {
+            titleTuple = new CheckboxTextfieldTuple(dialogContent, titleCheckboxSelector, titleTextfieldSelector);
+            descriptionTuple = new CheckboxTextfieldTuple(dialogContent, descriptionCheckboxSelector, descriptionTextfieldSelector, true);
+            $cqFileUpload     = $dialog.find(".cq-FileUpload");
```
Comment (jckautzmann) [inline content/src/content/jcr_root/apps/core/wcm/components/download/v1/download/clientlibs/editor/js/download.js]: I'd define constants for all the strings:
".cq-FileUpload", ".cq-FileUpload-edit", "assetselected", "[coral-fileupload-clear]", "cqFileuploadFilereference", "/_jcr_content/metadata.json", "dc:description", "dc:title"

**Comment by jckautzmann** [comment]
@bkheadwire Thx for your PR! I went through the code and added a few comments. I installed it on a local instance: it looks good.
When playing with the component, I noticed an issue when scrolling the Title Type dropdown of the policy dialog.
I only tested with a .jpg asset from DAM. Does the component support other mime types? Videos, documents (.doc, .pdf)?
This component still needs to support defining file name, file format, file size: are you intending to add this to the PR?

File: content/src/content/jcr_root/apps/core/wcm/components/download/v1/download/download.html [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+    <div class="cmp-download__image" data-sly-test.imagePath="${download.imagePath}">
+        <img src="${imagePath}"/>
+    </div>
+    <h2 class="cmp-download__title" data-sly-test.title="${download.title}" data-sly-element="${download.titleType}">
+        ${title}
+    </h2>
+    <div class="cmp-download__description" data-sly-test.description="${download.description}">
+        ${description @ context="html"}
+    </div>
+    <a class="cmp-download__cta" href="${download.downloadUrl}">${download.ctaText}</a>
```
Comment (richardhand) [inline content/src/content/jcr_root/apps/core/wcm/components/download/v1/download/download.html]: In the teaser component we refer to CTAs as "action" in code and "Call-To-Action" in the dialog - we should be consistent here.

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Download.java (hunk 4 above)
Comment (richardhand) [reply bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Download.java]: @jckautzmann - I think it's clear enough, and it follows our naming in the teaser [0]. What we are missing here are some docs for these properties as we have in the other component interfaces.

[0] https://github.com/adobe/aem-core-wcm-components/blob/master/bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Teaser.java#L109

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Download.java [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+     */
+    default String getTitle(){
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * Returns either the description configured in the dialog or the title of the DAM asset,
+     * depending on the state of the descriptionFromAsset checkbox.
+     *
+     * @return the download title
```
Comment (richardhand) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Download.java]: "the download description"

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Download.java [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+     * depending on the state of the titleFromAsset checkbox.
+     *
+     * @return the download title
+     */
+    default String getTitle(){
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * Returns either the description configured in the dialog or the title of the DAM asset,
```
Comment (richardhand) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Download.java]: "description of the DAM asset"

Although I'm also not sure we should reveal the implementation details in these interface descriptions (dialog fields, types etc.) - I also would be inclined not to mention how the title is retrieved, which may (although unlikely) be subject to change in future version implementations. Same comment for the description.

**Comment by bkheadwire** [comment]
> @bkheadwire Thx for your PR! I went through the code and added a few comments. I installed it on a local instance: it looks good.

Thank you for all of the detailed feedback! I've addressed many of the items you mentioned and will continue to do so as I have time.

> When playing with the component, I noticed an issue when scrolling the Title Type dropdown of the policy dialog.

Yep, I see that. It looks like the entire panel is trying to scroll in addition to the dropdown itself. Haven't had a chance to look at that yet.

> I only tested with a .jpg asset from DAM. Does the component support other mime types? Videos, documents (.doc, .pdf)?

When the asset is dropped from the DAM, every format I've tried has at least been downloadable (which isn't particularly surprising - currently the download link points straight at the asset and exposes the DAM structure which is something I've gotten some feedback from @vladbailescu about changing). However, if the asset doesn't have a usable image rendition, the thumbnail will be broken. There's also currently a problem with the dialog not properly loading the thumbnail or the properties from the DAM after it's been closed and reopened.

The browse functionality currently doesn't work at all. :-1: 

> This component still needs to support defining file name, file format, file size (according to the requirements in #217): are you intending to add this to the PR?

Yes!

**Comment by kwin** [comment]
Currently this PR does not set a content disposition header (which would require a dedicated servlet, compare with https://github.com/adobe/aem-core-wcm-components/issues/217#issuecomment-402534042). Also in some customer setups the direct access to DAM paths/servlets is prohibited by the dispatcher.

**Comment by vladbailescu** [comment]
Mergin this PR into the `feature/download` branch for now. Will add improvements on top of that.

**Pull request merged**